xblock/scorable.py

```diff
@@ -0,0 +1,140 @@
+"""
+Scorable.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from collections import namedtuple
+import logging
+
+
+log = logging.getLogger(__name__)
+
+
+Score = namedtuple('Score', ['earned', 'total'])
+
+
+class ScorableXBlockMixin(object):
+    """
+    Mixin to handle functionality related to scoring.
+
+    Provides the following public methods:
+
+        ScorableXBlockMixin.rescore(self, only_if_higher)
+            Calculate a new score and save it to the block.
+            `only_if_higher
+
+        ScorableXBlockMixin.allows_rescore(self) -> bool
+            Returns True if the block can be rescored.
+
+    Subclasses must define the following:
+
+        ScorableXBlockMixin.get_score(self)
+            Return the currently saved score.  Must not calculate a new score.
+
+        ScorableXBlockMixin.set_score(self, score)
+            Set the score on the block to the specified value
+
+        ScorableXBlockMixin.calculate_score(self, score)
+            Calculate a new score based on the state of the block.  Must not
+            update any values on the block.
+    """
+
+    def rescore(self, only_if_higher):
+        """
+        Calculate a new raw score and save it to the block.  If only_if_higher
+        is True and the score didn't improve, keep the existing score.
+
+        Raises a TypeError if the block cannot be scored.
+        Raises a ValueError if the user has not yet completed the problem.
+        """
+
+        _ = self.runtime.service(self, 'i18n').ugettext
+        errorlog_prefix = _('Rescore failure:')
```

Comment:
The `log` statements in this method seem superfluous and will cause duplication of logs.  It's sufficient to simply raise the error with the correct message and let higher callers log.  See the caller's implementation in https://github.com/edx/edx-platform/blob/master/lms/djangoapps/instructor_task/tasks_helper.py#L479